docs(clients): updates the tuning content for producers and consumers (#9260)
strimzi/strimzi-kafka-operator: documentation/modules/managing/con-consumer-config-properties.adoc

= Kafka consumer configuration tuning

[role="_abstract"]
Use configuration properties to optimize the performance of Kafka consumers.

> You can use standard Kafka consumer configuration options.

Reviewer: TBH, this sentence seems weird. What are the `standard options`? Is this expected to refer to the options that Kafka supports? Or some subset of important or common options? I would probably leave this out.

When tuning your consumers your primary concern will be ensuring that they cope efficiently with the amount of data ingested.
As with the producer tuning, be prepared to make incremental changes until the consumers operate as expected.

When tuning a consumer, consider the following aspects carefully, as they significantly impact its performance and behavior:

> Scaling:: Consumer groups enable parallel processing of messages by distributing the load across multiple consumers. Properly configuring consumer groups can enhance scalability and throughput.

Reviewer: I think this needs to mention partitions. Without them, this is confusing as users might not understand how the consumer groups scale.

Reviewer 2: Agree with @scholzj. We should say that the number of topic partitions determines the max level of parallelism that you can achieve, because one partition can only be assigned to one consumer in a consumer group. That said, a single consumer can offload message processing to a thread pool when ordering is not mandatory, but this is not supported by the Kafka library.

> Message ordering:: A consumer observes messages in a single partition in the same order that they were committed to the broker, which means that Kafka only provides ordering guarantees for messages in a single partition. If absolute ordering within a topic is important, use a single-partition topic. While a single-partition topic allows only one consumer instance, you can leverage a pool of worker threads to create a multi-threaded consumer for more efficient processing. When processing messages for specific entities (like users), you can use the ID (like user ID) as the message key and direct all messages to a single partition. This way, you can maintain message ordering for events specific to individual entities. If a new entity is created, you can create a new topic and migrate its data.

Reviewer: This is super hard to comment on because it is one long line. I would switch the first two sentences. Single partition topic allows only `one consumer instance to consumer from it at a point in time`. But you can have more of them. The threading suggestion seems wrong to me => it makes it sound super easy, but without individual-per-message acks it is super hard to implement. The users as the keys are also a bit oversimplified.

Author: - Switched the first two lines
- I think we can live without the threading part here -- removed
- Not sure how much more in-depth we should get for using keys here. I changed the text a bit here

> Message ordering:: 
> If absolute ordering within a topic is important, use a single-partition topic. 
> A consumer observes messages in a single partition in the same order that they were committed to the broker, which means that Kafka only provides ordering guarantees for messages in a single partition. 
> It is also possible to maintain message ordering for events specific to individual entities, such as users.
> If a new entity is created, you can create a new topic dedicated to that entity.
> You can use a unique ID, like a user ID, as the message key and route all messages with the same key to a single partition within the topic.

> Offset reset policy:: Setting the appropriate offset policy ensures that the consumer consumes messages from the desired starting point and handles message processing accordingly. The default Kafka reset value is `latest`, which starts at the end of the partition, and consequently means some messages might be missed, depending on the consumer's behavior and the state of the partition. Setting `auto.offset.reset` to `earliest` ensures that when connecting with a `new group.id`, all messages are retrieved from the beginning of the log.

Reviewer: new `groups.id` instead of `new group.id`.

----
<1> Add a consumer to a consumer group using a group id.

== Choosing the right partition assignment strategy

> Select an appropriate partition assignment strategy, which determines how Kafka topic partitions are distributed among consumer instances in a group.

Reviewer: Should the actual classes be mentioned here by name? Without that nobody can configure it to anything.

Specify the strategy using the `partition.assignment.strategy` consumer configuration property. 
The *range* assignment strategy assigns a range of partitions to each consumer, and is useful when you want to process related data together.

Alternatively, opt for a *round robin* assignment strategy for equal partition distribution among consumers, which is ideal for high-throughput scenarios requiring parallel processing.

For more stable partition assignments, consider the *sticky* and *cooperative sticky* strategies. 
Sticky strategies maintain assigned partitions during rebalances, when possible. 

> The cooperative sticky strategy also supports cooperative rebalances, enabling uninterrupted consumption from partitions that are not reassigned.

Reviewer: Based on what do they maintain the stickiness?

Author: I updated to mention that stickiness is based on previously assigned partitions.

> For more stable partition assignments, consider the *sticky* and *cooperative sticky* strategies. 
> Sticky strategies maintain assigned partitions during rebalances, when possible. 
> If a consumer was previously assigned certain partitions, the sticky strategies aim to reassign those same partitions back to the same consumer after a rebalance.
> Leaving partition assignments in place reduces the overhead on partition movements.
> The cooperative sticky strategy also supports cooperative rebalances, enabling uninterrupted consumption from partitions that are not reassigned.

----
<1> Add a consumer to a consumer group using a group id.

== Choosing the right partition assignment strategy

> Select an appropriate partition assignment strategy, which determines how Kafka topic partitions are distributed among consumer instances in a group.

Reviewer 3: @showuon Should we recommend users to set the strategy to `CooperativeStickyAssignor`? I remember a few issues with it, have they been all solved? 
By default consumers still use `RangeAssignor` which in many cases is literally the worst option.

Reviewer 4: Yes, AFAIK, there's a bug related to CooperativeStickyAssignor that could cause duplicated consumption still open: https://issues.apache.org/jira/browse/KAFKA-14224 . But yes, rangeAssignor is not a good default one, but I think that's a safe one. Here, I think it's better we don't recommend any assignor and let users to choose it.

Author: Thanks. We don't recommend a strategy and also suggest that a custom strategy might work better, so we can leave the section as it is.

----
<1> Add a consumer to a consumer group using a group id.

== Choosing the right partition assignment strategy

Select an appropriate partition assignment strategy, which determines how Kafka topic partitions are distributed among consumer instances in a group.

Partition strategies are supported by the following classes:

* `org.apache.kafka.clients.consumer.RangeAssignor`
* `org.apache.kafka.clients.consumer.RoundRobinAssignor`
* `org.apache.kafka.clients.consumer.StickyAssignor`
* `org.apache.kafka.clients.consumer.CooperativeStickyAssignor`

Specify a class using the `partition.assignment.strategy` consumer configuration property. 
The *range* assignment strategy assigns a range of partitions to each consumer, and is useful when you want to process related data together.

Alternatively, opt for a *round robin* assignment strategy for equal partition distribution among consumers, which is ideal for high-throughput scenarios requiring parallel processing.

For more stable partition assignments, consider the *sticky* and *cooperative sticky* strategies. 
Sticky strategies maintain assigned partitions during rebalances, when possible. 

> If a consumer was previously assigned certain partitions, the sticky strategies aim to reassign those same partitions back to the same consumer after a rebalance.

Reviewer 5: IIRC it's not going to "reassign" in the sense that it revokes and reassigns, leaving the consumer not consuming until the reassignment happens. IIRC it doesn't revoke the partitions which should stick with that consumer. Just revokes the ones which will be moved to another consumer. @mimaison can you confirma that? If yes, @PaulRMellor could we make it clearer here?

Author: That's a fair point. I don't think reassign is a good way of wording this. I've updated here:  [docs(review): edits to doc from review by PP](https://github.com/strimzi/strimzi-kafka-operator/pull/9260/commits/f8c19a7ce3d24e6f61064f3a81b073034c3883b1)

Reviewer: You fine with this @ppatierno? Can we merge it?

Reviewer 5: Yes it's fine with me. Thanks.

== Recovering from failure to avoid data loss

> Kafka provides mechanisms to detect and recover from failures within consumer groups. Two useful properties that aid in this recovery process are `session.timeout.ms` and `heartbeat.interval.ms`.

Reviewer 2: ```suggestion
Kafka provides a rebalance protocol which helps with failure detection and recovery within a consumer group. Two useful properties that aid in this recovery process are `session.timeout.ms` and `heartbeat.interval.ms`.
```

== Recovering from failure to avoid data loss

Kafka provides a rebalance protocol which helps with failure detection and recovery within a consumer group.  

> Two useful properties that aid in this recovery process are `session.timeout.ms` and `heartbeat.interval.ms`.

Reviewer 3: In most cases users should not have to tweak these settings. They can be used for relatively advanced use cases if there's a good justification, in my experience tweaking these early is often a red flag and instead tuning `max.poll.records` is preferred.

Author: I'v changed the priority of this section to put the emphasis on `max.poll.records`
